Cucumber Exhaustive Combination: inSet array containing invalid number (+12) does not throw an error.

## Bug Report
A test containing an invalid numeric value (positively signed) within an array does not throw an error as expected.

> Scenario: Running an exhaustive combination strategy with invalid number values should fail with an appropriate error message
>      Given the following fields exist:
>        | foo |
>        | bar |
>        And foo is of type "numeric"
>        And bar is of type "numeric"
>        And foo is in set:
>          | 999 |
>          **| +12 |**
>        And bar is in set:
>          | 12.01 |
>          | 0     |
>      Then I am presented with an error message
>        And no data is created

### Steps to Reproduce:
1. Run tests from the feature file ExhaustiveCombination.feature
2. See that there is a failure for the test **Running an exhaustive combination strategy with invalid number values should fail with an appropriate error message**
3. Note that the error message asserts that no failure is generated.

### Expected Result:
Adding an invalid numeric value of | +12 | for a field specified as numeric should result in an error.

### Actual Result:
The expected error was not generated, resulting in the test failing with an error message:

> java.lang.AssertionError: 
> Expected: is `true`
>      but: was `false`
> 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
> 	at org.junit.Assert.assertThat(Assert.java:956)
> 	at org.junit.Assert.assertThat(Assert.java:923)
> 	at com.scottlogic.deg.generator.cucumber.steps.GeneralTestStep.dataGeneratorShouldError(GeneralTestStep.java:88)
> 	at ✽.I am presented with an error message(C:/DataGenerator/data-engineering-generator/generator/src/test/java/com/scottlogic/deg/generator/cucumber/ExhaustiveCombination.feature:133)

### Environment:
ExhaustiveCombination branch running in full generation strategy & exhaustive combination strategy on IntelliJ.

### Additional Context:
![image](https://user-images.githubusercontent.com/43746958/49524581-f2b36180-f8a3-11e8-803d-4c7fce39b05c.png)

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/EqualTo.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/EqualTo.feature
@@ -299,9 +299,11 @@ Scenario: Running an 'equalTo' request that includes a negative number value sho
        | null |
        | -1   |
 
+@ignore #Ignored as leading zeros are not supported by jackson when parsing/reading JSON.
 Scenario: Running an 'equalTo' request that includes a number with Preceding zeros should be unsuccessful
      Given there is a field foo
-       But the profile is invalid as foo can't be equal to 001
+       And foo is equal to 001
+     Then the profile is invalid
 
 Scenario: Running an 'equalTo' request that includes a decimal number with trailing zeros should be successful
      Given there is a field foo

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GranularTo.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GranularTo.feature
@@ -94,7 +94,7 @@ Scenario: User requires to create a numeric field with negative data values that
        | -1.0 |
 
 Scenario: User attempts to create a numeric field with data value that include a decimal value to one decimal point incorrectly using a string to set the granularity
-     Given there is a field foo
-       But the profile is invalid as foo can't be granular to "0.1"
-     Then I am presented with an error message
+  Given there is a field foo
+  And foo is granular to "0.1"
+     Then the profile is invalid
         And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GreaterThan.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GreaterThan.feature
@@ -115,6 +115,6 @@ Scenario: User requires to create a numeric field with data values that are grea
 
 Scenario: User attempts to create a numeric field with data value that are greater than zero using an incorrect field value type of string
      Given there is a field foo
-       But the profile is invalid as foo can't be greater than "Zero"
-     Then I am presented with an error message
+       And foo is greater than "Zero"
+     Then the profile is invalid
         And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GreaterThanOrEqualTo.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/GreaterThanOrEqualTo.feature
@@ -167,6 +167,6 @@ Scenario: User requires to create a numeric field with data values that are grea
 
 Scenario: User attempts to create a numeric field with data value that are greater than or the same as zero using an incorrect field value type of string
      Given there is a field foo
-       But the profile is invalid as foo can't be greater than or equal to "Zero"
-     Then I am presented with an error message
+       And foo is greater than or equal to "Zero"
+     Then the profile is invalid
         And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/LessThan.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/LessThan.feature
@@ -95,6 +95,6 @@ Scenario: User requires to create a numeric field with data values that are less
 
 Scenario: User attempts to create a numeric field with data value that are less than zero using an incorrect field value type of string
      Given there is a field foo
-       But the profile is invalid as foo can't be less than "Zero"
-     Then I am presented with an error message
+       And foo is less than "Zero"
+     Then the profile is invalid
         And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/LessThanOrEqualTo.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/LessThanOrEqualTo.feature
@@ -142,6 +142,6 @@ Scenario: User requires to create a numeric field with data values that are less
 
 Scenario: User attempts to create a numeric field with data value that are less than or the same as zero using an incorrect field value type of string
      Given there is a field foo
-       But the profile is invalid as foo can't be less than or equal to "Zero"
-     Then I am presented with an error message
+       And foo is less than or equal to "Zero"
+     Then the profile is invalid
         And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/BeforeInteresting.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/BeforeInteresting.feature
@@ -170,34 +170,34 @@ Scenario: User creates data that is anything but before a specified date
 @ignore
 Scenario: User attempts to create data before a specified invalidly formatted date
      Given there is a field foo
-       But the profile is invalid as foo can't be before "2018-10-01"
-     Then I am presented with an error message
+       And foo is before "2018-10-01"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before a specified invalid date
      Given there is a field foo
-       But the profile is invalid as foo can't be before 2018-10-40T00:00:00.000
-     Then I am presented with an error message
+       And foo is before 2018-10-40T00:00:00.000
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before a specified invalidly formatted time
      Given there is a field foo
-       But the profile is invalid as foo can't be before "2018-10-01T01:00:00.000AM"
-     Then I am presented with an error message
+       And foo is before "2018-10-01T01:00:00.000AM"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before a specified invalid time
      Given there is a field foo
-       But the profile is invalid as foo can't be before 2018-10-01T50:00:00.000
-     Then I am presented with an error message
+       And foo is before 2018-10-01T50:00:00.000
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before a specified date and an invalidly formatted date
      Given there is a field foo
-     But the profile is invalid as foo can't be before "2018-10-01"
-       Then I am presented with an error message
-     And no data is created
+       And foo is before "2018-10-01"
+     Then the profile is invalid
+       And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/BeforeOrAtInteresting.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/BeforeOrAtInteresting.feature
@@ -172,35 +172,35 @@ Scenario: User creates data that is anything but before or equal to a specified 
 @ignore
 Scenario: User attempts to create data before or equal to a specified invalidly formatted date
      Given there is a field foo
-       But the profile is invalid as foo can't be before or at "2018-10-01"
-     Then I am presented with an error message
+       And foo is before or at "2018-10-01"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before or equal to a specified invalid date
      Given there is a field foo
-       But the profile is invalid as foo can't be before or at 2018-10-40T00:00:00.000
-     Then I am presented with an error message
+       And foo is before or at 2018-10-40T00:00:00.000
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before or equal to a specified invalidly formatted time
      Given there is a field foo
-       But the profile is invalid as foo can't be before or at "2018-10-01T01:00:00.000AM"
-     Then I am presented with an error message
+       And foo is before or at "2018-10-01T01:00:00.000AM"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before or equal to a specified invalid time
      Given there is a field foo
-       But the profile is invalid as foo can't be before or at 2018-10-01T50:00:00.000
-     Then I am presented with an error message
+       And foo is before or at 2018-10-01T50:00:00.000
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to create data before or equal to a specified date and an invalidly formatted date
      Given there is a field foo
        And foo is before 2018-10-10T00:00:00.000
-       But the profile is invalid as foo can't be before or at "2018-10-01"
-     Then I am presented with an error message
+       And foo is before or at "2018-10-01"
+     Then the profile is invalid
        And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/GreaterThanInteresting.feature
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/interesting/GreaterThanInteresting.feature
@@ -276,20 +276,20 @@ Scenario: User creates data that is anything but higher than a specified number
 @ignore
 Scenario: User attempts to data higher than a specified string
      Given there is a field foo
-       But the profile is invalid as foo can't be greater than "$10.00"
-     Then I am presented with an error message
+       And foo is greater than "$10.00"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to data higher than a specified number formatted as a string
      Given there is a field foo
-       But the profile is invalid as foo can't be greater than "1"
-     Then I am presented with an error message
+       And foo is greater than "1"
+     Then the profile is invalid
        And no data is created
 
 @ignore
 Scenario: User attempts to data higher than a specified date
      Given there is a field foo
-       But the profile is invalid as foo can't be greater than 2018-10-10T00:00:00.000
-     Then I am presented with an error message
+       And foo is greater than 2018-10-10T00:00:00.000
+     Then the profile is invalid
        And no data is created

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/DateObject.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/DateObject.java
@@ -1,0 +1,9 @@
+package com.scottlogic.deg.generator.cucumber.steps;
+
+import java.util.HashMap;
+
+public class DateObject extends HashMap {
+    public DateObject(String date){
+        this.put("date", date);
+    }
+}

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/DateValueStep.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/DateValueStep.java
@@ -15,19 +15,14 @@ public class DateValueStep {
         this.state = state;
     }
 
-    @When("{fieldVar} is {operator} {dateString}")
-    public void whenFieldIsConstrainedByDateValue(String fieldName, String constraintName, String value) throws Exception {
-        this.state.addConstraint(fieldName, constraintName, dateObject(value));
+    @When("{fieldVar} is {operator} {date}")
+    public void whenFieldIsConstrainedByDateValue(String fieldName, String constraintName, DateObject value) throws Exception {
+        this.state.addConstraint(fieldName, constraintName, value);
     }
 
-    @When("{fieldVar} is anything but {operator} {dateString}")
-    public void whenFieldIsNotConstrainedByDateValue(String fieldName, String constraintName, String value) throws Exception {
-        this.state.addNotConstraint(fieldName, constraintName, dateObject(value));
-    }
-
-    public static Map dateObject(String dateValue) {
-        HashMap map = new HashMap();
-        map.put("date", dateValue);
-        return map;
+    @When("{fieldVar} is anything but {operator} {date}")
+    public void whenFieldIsNotConstrainedByDateValue(String fieldName, String constraintName, DateObject value) throws Exception {
+        this.state.addNotConstraint(fieldName, constraintName, value);
     }
 }
+

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/GeneralTestStep.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/GeneralTestStep.java
@@ -61,15 +61,8 @@ public class GeneralTestStep {
         this.state.addNotConstraint(fieldName, "null", null);
     }
 
-    @But("^the profile is invalid as (.+) can't be ([a-z ]+) (((\".*\")|(" + DateValueStep.DATE_REGEX + ")|([0-9]+(.[0-9]+){1}))+)$")
-    public void fieldIsInvalid(String fieldName, String constraint, String value) {
-        try {
-            Object parsedValue = GeneratorTestUtilities.parseInput(value);
-            this.state.addConstraint(fieldName, constraint, parsedValue);
-        } catch (Exception e) {
-            this.state.addException(e);
-        }
-
+    @Then("^the profile is invalid$")
+    public void theProfileIsInvalid() {
         testHelper.generateAndGetData();
 
         Assert.assertThat(
@@ -101,6 +94,7 @@ public class GeneralTestStep {
     @Then("^I am presented with an error message$")
     public void dataGeneratorShouldError() {
         testHelper.generateAndGetData();
+
         Assert.assertThat(testHelper.generatorHasThrownException(), is(true));
     }
 

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/SetValueStep.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/SetValueStep.java
@@ -1,14 +1,10 @@
 package com.scottlogic.deg.generator.cucumber.steps;
 
-import com.fasterxml.jackson.core.JsonParseException;
 import com.scottlogic.deg.generator.cucumber.utils.DegTestState;
-import com.scottlogic.deg.generator.cucumber.utils.GeneratorTestUtilities;
-import com.scottlogic.deg.generator.inputs.InvalidProfileException;
 import cucumber.api.java.en.When;
 
-import java.util.Collection;
 import java.util.List;
-import java.util.stream.Collectors;
+import java.util.Set;
 
 public class SetValueStep {
 
@@ -19,27 +15,12 @@ public class SetValueStep {
     }
 
     @When("{fieldVar} is in set:")
-    public void whenFieldIsConstrainedBySetValue(String fieldName, List<String> values) {
-        this.state.addConstraint(fieldName, "in set", this.getSetValues(values));
+    public void whenFieldIsConstrainedBySetValue(String fieldName, List<Object> values) {
+        this.state.addConstraint(fieldName, "in set", values);
     }
 
     @When("{fieldVar} is anything but in set:")
-    public void whenFieldIsNotConstrainedBySetValue(String fieldName, List<String> values) {
-        this.state.addNotConstraint(fieldName, "in set", this.getSetValues(values));
+    public void whenFieldIsNotConstrainedBySetValue(String fieldName, List<Object> values) {
+        this.state.addNotConstraint(fieldName, "in set", values);
     }
-
-    private Collection<Object> getSetValues(List<String> values) {
-        return values.stream()
-            .map(String::trim)
-            .map(value -> {
-                try {
-                    return GeneratorTestUtilities.parseInput(value);
-                } catch (JsonParseException | InvalidProfileException e) {
-                    this.state.addException(e);
-                    return "<exception thrown: " + e.getMessage() + ">";
-                }
-            })
-            .collect(Collectors.toSet());
-    }
-
 }

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
@@ -1,12 +1,18 @@
 package com.scottlogic.deg.generator.cucumber.steps;
 
+import com.fasterxml.jackson.core.JsonParseException;
+import com.scottlogic.deg.generator.cucumber.utils.DegTestState;
 import com.scottlogic.deg.generator.cucumber.utils.GeneratorTestUtilities;
 import com.scottlogic.deg.generator.generation.GenerationConfig;
+import com.scottlogic.deg.generator.inputs.InvalidProfileException;
 import com.scottlogic.deg.schemas.v3.AtomicConstraintType;
+import cucumber.api.DataTable;
 import cucumber.api.TypeRegistry;
 import cucumber.api.TypeRegistryConfigurer;
 import io.cucumber.cucumberexpressions.ParameterType;
 import io.cucumber.cucumberexpressions.Transformer;
+import io.cucumber.datatable.DataTableType;
+import io.cucumber.datatable.TableCellByTypeTransformer;
 
 import java.util.*;
 import java.util.stream.Collectors;
@@ -27,6 +33,7 @@ public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
         this.defineOperationParameterType(tr);
         this.defineParameterType(tr,"fieldVar", "^(.+)");
         this.defineParameterType(tr,"regex", "/(.+)/$");
+        tr.setDefaultDataTableCellTransformer(new DataTableCellTransformer());
 
         tr.defineParameterType(new ParameterType<>(
             "number",
@@ -106,4 +113,14 @@ public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
             .collect(Collectors.joining("|", "(", ")"));
     }
 
+    private class DataTableCellTransformer implements TableCellByTypeTransformer {
+        @Override
+        public <T> T transform(String value, Class<T> aClass) throws Throwable {
+            try {
+                return aClass.cast(GeneratorTestUtilities.parseInput(value.trim()));
+            } catch (JsonParseException | InvalidProfileException e) {
+                return (T)e;
+            }
+        }
+    }
 }

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
@@ -37,7 +37,7 @@ public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
 
         tr.defineParameterType(new ParameterType<>(
             "number",
-            "(-?\\d+(\\.\\d+)?)$",
+            "([+-]?\\d+(\\.\\d+)?)$",
             Number.class,
             (Transformer<Number>) value -> (Number) GeneratorTestUtilities.parseNumber(value)));
 

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/steps/TypeRegistryConfiguration.java
@@ -26,7 +26,6 @@ public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
         this.defineCombinationStrategyType(tr);
         this.defineOperationParameterType(tr);
         this.defineParameterType(tr,"fieldVar", "^(.+)");
-        this.defineParameterType(tr,"dateString", DateValueStep.DATE_REGEX);
         this.defineParameterType(tr,"regex", "/(.+)/$");
 
         tr.defineParameterType(new ParameterType<>(
@@ -40,6 +39,12 @@ public class TypeRegistryConfiguration implements TypeRegistryConfigurer {
             "(true|false)$",
             Boolean.class,
             (Transformer<Boolean>) Boolean::valueOf));
+
+        tr.defineParameterType(new ParameterType<>(
+            "date",
+            DateValueStep.DATE_REGEX,
+            DateObject.class,
+            DateObject::new));
     }
 
     private void defineOperationParameterType(TypeRegistry tr){

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/DegTestState.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/DegTestState.java
@@ -1,16 +1,15 @@
 package com.scottlogic.deg.generator.cucumber.utils;
 
+import com.fasterxml.jackson.core.JsonParseException;
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.scottlogic.deg.generator.Field;
 import com.scottlogic.deg.generator.generation.GenerationConfig;
+import com.scottlogic.deg.generator.inputs.InvalidProfileException;
 import com.scottlogic.deg.schemas.v3.ConstraintDTO;
 
 import java.io.IOException;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.List;
+import java.util.*;
 import java.util.stream.Collectors;
 
 public class DegTestState {
@@ -22,6 +21,20 @@ public class DegTestState {
     final List<Field> profileFields = new ArrayList<>();
     final List<ConstraintDTO> constraints = new ArrayList<>();
     final List<Exception> testExceptions = new ArrayList<>();
+
+    public void addConstraint(String fieldName, String constraintName, List<Object> value) {
+        if (value == null)
+            addConstraint(fieldName, constraintName, (Object)value);
+        else
+            addConstraint(fieldName, constraintName, getSetValues(value));
+    }
+
+    public void addNotConstraint(String fieldName, String constraintName, List<Object> value) {
+        if (value == null)
+            addNotConstraint(fieldName, constraintName, (Object)value);
+        else
+            addNotConstraint(fieldName, constraintName, getSetValues(value));
+    }
 
     public void addConstraint(String fieldName, String constraintName, Object value) {
         ConstraintDTO dto = this.createConstraint(fieldName, constraintName, value);
@@ -37,6 +50,21 @@ public class DegTestState {
     public void addConstraintsFromJson(String constraintProfile) throws IOException {
         ConstraintHolder holder = this.deserialise(constraintProfile);
         this.constraints.addAll(holder.constraints);
+    }
+
+    private Collection<Object> getSetValues(List<Object> values) {
+        if (values == null){
+            throw new IllegalArgumentException("Values cannot be null");
+        }
+
+        values.stream()
+            .filter(value -> value instanceof Exception)
+            .map(value -> (Exception)value)
+            .forEach(this::addException);
+
+        return values.stream()
+            .filter(value -> !(value instanceof Exception))
+            .collect(Collectors.toSet());
     }
 
     public void clearState(){

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/GeneratorTestUtilities.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/GeneratorTestUtilities.java
@@ -8,6 +8,7 @@ import com.scottlogic.deg.generator.Profile;
 import com.scottlogic.deg.generator.ProfileFields;
 import com.scottlogic.deg.generator.Rule;
 import com.scottlogic.deg.generator.constraints.Constraint;
+import com.scottlogic.deg.generator.cucumber.steps.DateObject;
 import com.scottlogic.deg.generator.cucumber.steps.DateValueStep;
 import com.scottlogic.deg.generator.decisiontree.DecisionTreeCollection;
 import com.scottlogic.deg.generator.decisiontree.ProfileDecisionTreeFactory;
@@ -109,7 +110,7 @@ public class GeneratorTestUtilities {
         if (input.startsWith("\"") && input.endsWith("\"")) {
             return input.substring(1, input.length() - 1);
         } else if (input.matches(DateValueStep.DATE_REGEX)) {
-            return DateValueStep.dateObject(input);
+            return new DateObject(input);
         } else if (input.equals("null")) {
             return null;
         } else if (input.matches("-?(\\d+(\\.\\d+)?)")) {

--- a/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/GeneratorTestUtilities.java
+++ b/generator/src/test/java/com/scottlogic/deg/generator/cucumber/utils/GeneratorTestUtilities.java
@@ -113,7 +113,7 @@ public class GeneratorTestUtilities {
             return new DateObject(input);
         } else if (input.equals("null")) {
             return null;
-        } else if (input.matches("-?(\\d+(\\.\\d+)?)")) {
+        } else if (input.matches("[+-]?(\\d+(\\.\\d+)?)")) {
             return parseNumber(input);
         } else if (input.equals("true") || input.equals("false")){
             return input.equals("true");

